language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -289,25 +309,30 @@ typeOf (TyInst x body ty) = do
             substituteNormalizeType vTy n vCod
         _ -> throwError (TypeMismatch x (void body) (TyForall () dummyTyName dummyKind dummyType) vBodyTy)
 
--- [infer| term : fix n vPat]    [fix n vPat / n] vPat ~> vRes
--- -----------------------------------------------------------
--- [infer| unwrap n term : vRes]
+-- [infer| term : ifix vPat vArg]    [infer| vArg :: k]
+-- ------------------------------------------------------------------
+-- [infer| unwrap term : NORM (vPat (\(a :: k) -> ifix vPat a) vArg)]
 typeOf (Unwrap x term) = do
     vTermTy <- typeOf term
     case getNormalizedType vTermTy of
-        TyFix _ n vPat -> substituteNormalizeType vTermTy n vPat
-        _              -> throwError (TypeMismatch x (void term) (TyFix () dummyTyName dummyType) vTermTy)
-
--- [check| pat :: *]    pat ~>? vPat    [fix n vPat / n] vPat ~> vTermTy'    [check| term : vTermTy]
--- -------------------------------------------------------------------------------------------------
--- [infer| wrap n pat term : fix n vPat]
-typeOf (Wrap x n pat term) = do
-    kindCheckM x pat $ Type ()
-    vPat <- normalizeTypeOpt $ void pat
-    vTermTy <- substituteNormalizeType (TyFix () (void n) <$> vPat) (void n) $ getNormalizedType vPat
+        TyIFix _ vPat vArg -> do
+            k <- kindOf $ x <$ vArg  -- Looks weird.
+            unfoldFixOf vPat vArg k
+        _                  -> throwError (TypeMismatch x (void term) (TyIFix () dummyType dummyType) vTermTy)
+
+-- [infer| pat :: (k -> *) -> k -> *]    [check | arg :: k]    pat ~>? vPat    arg ~>? vArg
+-- [check| term : NORM (vPat (\(a :: k) -> ifix vPat a) vArg)]
+-- ----------------------------------------------------------------------------------------
+-- [infer| iwrap pat arg term : ifix vPat vArg]
+typeOf (IWrap x pat arg term) = do
+    k <- indexOfPatternFunctor x pat
+    kindCheckM x arg k
+    vPat <- normalizeType {- Opt -} $ void pat
```

Comment:
Should this be `normalizeTypeOpt`?

Reply:
It's not clear. We generate 2-ary pattern functors that are not normalized even if you supply a normalized (n+3)-ary pattern functors. I do not know at the moment how to handle this.

Reply:
I'll trying to compute some stuff on the Haskell side rather than on the PLC side, that should give us more reduced types.

Reply:
I tried. Got more reduced types, but still not normal. This is a non-trivial problem.

Reply:
The answer: if you want to type check your code with type annotations normalization disabled, you better normalize your types first. Hence we can call `normalizeTypeOpt` there and the problem is resolved.